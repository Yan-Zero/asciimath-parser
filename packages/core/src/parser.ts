import { TokenTypes } from './symbols'
import type { TokenizedValue } from './trie'

enum NodeTypes {
  Root = 'Root',
  Const = 'Const',
  ParamOne = 'ParamOne', // expression with one param
  ParamTwo = 'ParamTwo', // expression with two params
  Matrix = 'Matrix',
  Flat = 'Flat',
}

enum AlignDirection {
  Left = 'Left',
  Center = 'Center',
}

interface Node {
  type: NodeTypes
}

interface ConstNode extends Node {
  type: NodeTypes.Const
  value: string
  tex: string
}

interface RootNode extends Node {
  type: NodeTypes.Root
  body: ChildNode[]
}

interface FlatNode extends Node {
  type: NodeTypes.Flat
  body: ChildNode[]
}

interface ParamOneNode extends Node {
  type: NodeTypes.ParamOne
  params: ChildNode
  tex: string
}

interface ParamTwoNode extends Node {
  type: NodeTypes.ParamTwo
  params: [ChildNode, ChildNode]
  tex: string
}

interface MatrixNode extends Node {
  type: NodeTypes.Matrix
  params: ChildNode[][]
  lparen: string
  rparen: string
  alignment: AlignDirection
  dividerIndices: number[]
}

type ChildNode = ConstNode | MatrixNode | ParamOneNode | ParamTwoNode | FlatNode

function createRootNode(): RootNode {
  return {
    type: NodeTypes.Root,
    body: [],
  }
}
function createConstNode(): ConstNode
function createConstNode(tex: string): ConstNode
function createConstNode(token: TokenizedValue): ConstNode
function createConstNode(arg?: TokenizedValue | string) {
  if (typeof arg === 'undefined') {
    return {
      type: NodeTypes.Const,
      value: '',
      tex: '',
    }
  }
  if (typeof arg === 'string') {
    return {
      type: NodeTypes.Const,
      value: arg,
      tex: arg,
    }
  }
  return {
    type: NodeTypes.Const,
    value: arg.value,
    tex: arg.type === TokenTypes.Text ? `\\text{${arg.tex}}` : arg.tex,
  }
}

function createFlatNode(): FlatNode {
  return {
    type: NodeTypes.Flat,
    body: [],
  }
}

function createMatrixNode(): MatrixNode {
  return {
    type: NodeTypes.Matrix,
    params: [],
    lparen: '.',
    rparen: '.',
    alignment: AlignDirection.Center,
    dividerIndices: [],
  }
}

function createParamOneNode(): ParamOneNode {
  return {
    type: NodeTypes.ParamOne,
    tex: '',
    params: createFlatNode(),
  }
}

function createParamTwoNode(): ParamTwoNode {
  return {
    type: NodeTypes.ParamTwo,
    tex: '',
    params: [createFlatNode(), createFlatNode()],
  }
}

function createParenOfFlatNodeFrom(token: TokenizedValue, left: boolean): ConstNode {
  return { type: NodeTypes.Const, value: token.value, tex: `\\${left ? 'left' : 'right'}${token.tex}` }
}

function readParenedExpression2(tokens: TokenizedValue[], current: number): {
  node: MatrixNode | FlatNode
  current: number
} {
  const { closingIndex, semiIndex } = findPairedClosingParen(current, tokens)
  if (closingIndex === -1) {
    return parenStartedNoClosingNode(tokens, current)
  }
  else {
    if (semiIndex === -1) {
      // process the expression as an array
      return parenedArrayNode(tokens, current, closingIndex)
    }
    else {
      // process the expression as a matrix
      return generateMatrixNode(tokens, current, closingIndex)
    }
  }
}

function generateMatrixNode(tokens: TokenizedValue[], current: number, end: number) {
  let token = tokens[current]
  const node = createMatrixNode()
  const dividerIndices = new Set<number>()
  node.lparen = `\\left${token.tex}`
  token = tokens[++current]
  let tempArr: ChildNode[] = []
  let tempNode: ChildNode | null = null
  // inside a matrix
  while (current < end) {
    if (token.type === TokenTypes.Split) {
      if (token.value === ',') {
        if (tempNode) {
          tempArr.push(tempNode)
          tempNode = null
        }
        else {
          tempArr.push(createConstNode())
        }
      }
      else if (token.value === ';') {
        if (tempNode) {
          tempArr.push(tempNode)
          tempNode = null
        }
        node.params.push(tempArr)
        tempArr = []
      }
      token = tokens[++current]
      continue
    }
    else if (token.type === TokenTypes.Paren /* && token.value === '|' */) {
      if (tempNode) {
        tempArr.push(tempNode)
        tempNode = null
      }
      dividerIndices.add(tempArr.length)
      token = tokens[++current]
      continue
    }
    tempNode = createFlatNode()
    token = tokens[current]
    while (current < end
      && token.type !== TokenTypes.Split
      && token.type !== TokenTypes.Paren) {
      const walkRes = walk(tokens, current)
      current = walkRes.current
      tempNode.body.push(walkRes.node)
      token = tokens[current]
    }
  }
  // for those cases that the matrix only contains one line
  // or the last line of the matrix does not contains a semicolon
  if (tempNode) {
    tempArr.push(tempNode)
    tempNode = null
  }
  if (tempArr.length > 0) {
    node.params.push(tempArr)
    tempArr = []
  }
  // set dividerIndices
  node.dividerIndices = Array.from(dividerIndices).sort((a, b) => a - b)
  // process the right paren
  token = tokens[current]
  if (current < tokens.length) {
    current++
    node.rparen = `\\right${token.tex}`
    if (token.value === ':}')
      node.alignment = AlignDirection.Left
  }
  // unreachable
  else {
    // no right paren
    node.rparen = '\\right.'
  }
  return { node, current }
}

class ParenError extends Error {}

function parenedArrayNode(tokens: TokenizedValue[], current: number, closingIndex: number) {
  let token = tokens[current]
  const node = createFlatNode()
  node.body.push(createParenOfFlatNodeFrom(token, true))
  current = readTokensToFlatNode(current + 1, closingIndex, tokens, node)
  if (current >= tokens.length)
    throw new ParenError(`Read index out of range, index: ${current}`)

  token = tokens[current]
  current++
  node.body.push(createParenOfFlatNodeFrom(token, false))
  if ((node.body[0] as ConstNode).value === '{:' && (node.body[node.body.length - 1] as ConstNode).value === ':}') {
    (node.body[0] as ConstNode).tex = '{';
    (node.body[node.body.length - 1] as ConstNode).tex = '}'
  }
  return { node, current }
}

function parenStartedNoClosingNode(tokens: TokenizedValue[], current: number) {
  const token = tokens[current]
  // just process the expression as flat one
  const node = createFlatNode()
  node.body.push({ type: NodeTypes.Const, value: token.value, tex: `\\left${token.tex}` } as ConstNode)
  current = readTokensToFlatNode(current + 1, tokens.length, tokens, node)
  // maybe it's better to add a hidden closing bracket
  node.body.push({ type: NodeTypes.Const, value: token.value, tex: '\\right.' } as ConstNode)
  return { node, current }
}

function findPairedClosingParen(current: number, tokens: TokenizedValue[]) {
  let semiIndex = -1
  let closingIndex = -1
  const stack: string[] = []
  // find the paired closing paren and whether there is any `;` within the parens
  for (let i = current + 1; i < tokens.length; i++) {
    if (tokens[i].type === TokenTypes.LParen) {
      stack.push('')
      continue
    }
    if (stack.length === 0) {
      if (tokens[i].value === ';') {
        if (semiIndex === -1)
          semiIndex = i
      }
      else if (tokens[i].type === TokenTypes.RParen) {
        if (closingIndex === -1)
          closingIndex = i
      }
      if (semiIndex !== -1 && closingIndex !== -1)
        break
    }
    else {
      if (tokens[i].type === TokenTypes.RParen)
        stack.pop()
    }
  }
  return { closingIndex, semiIndex }
}

function findPairedBar(arr: TokenizedValue[], start: number, end: number): {
  semiIndex: number
  barIndex: number
} {
  let semiIndex = -1
  let barIndex = -1
  const stack: string[] = []
  for (let i = start; i < end; i++) {
    if (arr[i].type === TokenTypes.LParen) {
      stack.push('')
      continue
    }
    if (stack.length > 0 && arr[i].type === TokenTypes.RParen) {
      stack.pop()
      continue
    }
    if (stack.length > 0)
      continue
    if (arr[i].type === TokenTypes.RParen)
      break
    if (arr[i].value === ';') {
      if (semiIndex === -1)
        semiIndex = i
    }
    else if (arr[i].value === '|') {
      if (barIndex === -1)
        barIndex = i
    }
    if (semiIndex !== -1 && barIndex !== -1)
      break
  }
  return { semiIndex, barIndex }
}

function readBarStartedExpressions(tokens: TokenizedValue[], current: number): {
  node: ChildNode
  current: number
} {
  let token = tokens[current]
  // in fact only `|` matches `TokenTypes.Paren`
  const { semiIndex, barIndex } = findPairedBar(tokens, current + 1, tokens.length)

  /**
     * case 1: not matrix
     * there isn't any `;` split pattern before a Paren or a RParen
     * for example `{ (x, y) | x^2 + y^2 <= 1 }`
     */
  if (barIndex === -1)
    return createSingleBarNode(current)

  // used as `abs`
  if (semiIndex === -1 || semiIndex > barIndex) {
    const node = createFlatNode()
    current++
    node.body.push(createConstNode('\\left|'))
    current = readTokensToFlatNode(current, barIndex, tokens, node)
    node.body.push(createConstNode('\\right|'))
    current = barIndex + 1
    return { current, node }
  }

  /**
     * case 2: matrix
     * there is one or more than one `;` before a Paren ~~or a RParen~~
     *
     * common cases:
     * `| 1, 2; |` or `| 1, 2; 3, 4 |`
     *
     * special case:
     * `| 1, 2, | 3, 4; 5, 6 |, 4; ... |`
     * maybe I should use a `stack` to store the parens
     * however, the case is too complex, so I won't implement it ¯\_(ツ)_/¯
     */
  const node = createMatrixNode()
  node.lparen = '\\left|'
  node.rparen = '\\right|'

  token = tokens[++current]
  let tempArr: ChildNode[] = []
  let tempNode: ChildNode | null = null
  while (current < barIndex) {
    if (token.type === TokenTypes.Split) {
      switch (token.value) {
        case ',': {
          if (tempNode) {
            tempArr.push(tempNode)
            tempNode = null
          }
          break
        }
        case ';': {
          if (tempNode) {
            // the ownership of `tempNode` is passed to `tempArr`
            tempArr.push(tempNode)
            tempNode = null
          }
          // the ownership of `tempArr` is passed to `node.params`
          node.params.push(tempArr)
          tempArr = []
          break
        }
      }
      token = tokens[++current]
      continue
    }
    tempNode = createFlatNode()
    token = tokens[current]
    while (current < barIndex
        && token.type !== TokenTypes.Split) {
      const walkRes = walk(tokens, current)
      current = walkRes.current
      tempNode.body.push(walkRes.node)
      token = tokens[current]
    }
  }
  if (tempNode) {
    tempArr.push(tempNode)
    tempNode = null
  }
  if (tempArr.length > 0) {
    node.params.push(tempArr)
    tempArr = []
  }
  current = barIndex + 1
  return { node, current }
}

function readTokensToFlatNode(current: number, end: number, tokens: TokenizedValue[], node: FlatNode) {
  while (current < end) {
    const walkRes = walk(tokens, current)
    current = walkRes.current
    node.body.push(walkRes.node)
  }
  return current
}

function createSingleBarNode(current: number): { node: ChildNode; current: number } {
  return {
    current: current + 1,
    node: {
      type: NodeTypes.Const,
      value: '|',
      tex: '\\mid',
    },
  }
}

function removeParenOfFlatExpr(node: FlatNode): FlatNode {
  const first = node.body[0]
  const last = node.body[node.body.length - 1]
  if (first.type === NodeTypes.Const
    && last.type === NodeTypes.Const
    && (first as ConstNode).value === '('
    && (last as ConstNode).value === ')') {
    node.body.pop()
    node.body.shift()
  }
  return node
}

function lookForwardOperatorOptionalTwoParams(tokens: TokenizedValue[], current: number, token: TokenizedValue) {
  let nextToken = tokens[current]
  let p1: ChildNode = createConstNode()
  let nextShouldBe = ''
  let p1Status = ''
  let p2Status = ''
  if (nextToken.value === '^' || nextToken.value === '_') {
    nextShouldBe = nextToken.value === '^' ? '_' : '^'
    p1Status = nextToken.value
    current++
    // do not look forward
    // do not receive other sup or sub in recursion
    const walkRes = walk(tokens, current, false)
    if (walkRes.node.type === NodeTypes.Flat)
      walkRes.node = removeParenOfFlatExpr(walkRes.node)
    p1 = walkRes.node
    current = walkRes.current
  }
  let p2: ChildNode = createConstNode()
  if (current < tokens.length) {
    nextToken = tokens[current]
    if (nextToken.value === nextShouldBe) {
      p2Status = nextToken.value
      current++
      // do not look forward
      // do not receive other sup or sub in recursion
      const walkRes = walk(tokens, current, false)
      if (walkRes.node.type === NodeTypes.Flat)
        walkRes.node = removeParenOfFlatExpr(walkRes.node)
      p2 = walkRes.node
      current = walkRes.current
    }
  }
  const node = createParamTwoNode()
  node.tex = token.tex
  node.params[0] = (() => {
    if (p1Status === '^')
      return p1
    if (p2Status === '^')
      return p2
    return createConstNode()
  })()
  node.params[1] = (() => {
    if (p1Status === '_')
      return p1
    if (p2Status === '_')
      return p2
    return createConstNode()
  })()
  return { node, current }
}

function preProcessOperatorSup(node: ChildNode, operator: TokenizedValue, tokens: TokenizedValue[], current: number): WalkResult {
  let newNode: FlatNode
  if (node.type === NodeTypes.Flat) {
    newNode = node
  }
  else {
    newNode = createFlatNode()
    newNode.body.push(node)
  }
  const supNode = createParamOneNode()
  supNode.tex = operator.tex
  // do not look forward
  // avoid receiving fractions or other nodes as sup node
  const walkRes = walk(tokens, current, false)
  current = walkRes.current
  if (walkRes.node.type === NodeTypes.Flat)
    walkRes.node = removeParenOfFlatExpr(walkRes.node)
  supNode.params = walkRes.node
  newNode.body.push(supNode)
  node = newNode
  return { node, current }
}

function preProcessOperatorAO(node: ChildNode, nextToken: TokenizedValue) {
  const newNode = createFlatNode()
  if (node.type === NodeTypes.Flat)
    newNode.body.push(...(node.body))

  else
    newNode.body.push(node)
  newNode.body.push(createConstNode(nextToken))
  node = newNode
  return node
}

function preProcessOperatorAOB(node: ChildNode, operator: TokenizedValue, tokens: TokenizedValue[], current: number): WalkResult {
  const newNode = createParamTwoNode()
  if (node.type === NodeTypes.Flat)
    node = removeParenOfFlatExpr(node)
  newNode.tex = operator.tex
  newNode.params[0] = node
  const walkRes = walk(tokens, current)
  current = walkRes.current
  if (walkRes.node.type === NodeTypes.Flat)
    walkRes.node = removeParenOfFlatExpr(walkRes.node)
  newNode.params[1] = walkRes.node
  node = newNode
  return { node, current }
}

function getParamOneNode(tokens: TokenizedValue[], current: number, lookForward: boolean): { node: ParamOneNode; current: number } {
  const node = createParamOneNode()
  const token = tokens[current]
  node.tex = token.tex
  current++
  const walkRes = walk(tokens, current, lookForward)
  current = walkRes.current
  if (walkRes.node.type === NodeTypes.Flat)
    walkRes.node = removeParenOfFlatExpr(walkRes.node)

  node.params = walkRes.node
  return { node, current }
}

interface WalkResult {
  node: ChildNode
  current: number
}

function walk(tokens: TokenizedValue[], current: number, watchNext = true): WalkResult {
  if (current >= tokens.length)
    return { node: createConstNode(), current }
  const token = tokens[current]
  let node: ChildNode
  switch (token.type) {
    case TokenTypes.Const:
    case TokenTypes.Text:
    case TokenTypes.NumberLiteral:
    case TokenTypes.StringLiteral: {
      current++
      node = createConstNode(token)
      break
    }
    case TokenTypes.LParen: {
      ({ node, current } = readParenedExpression2(tokens, current))
      break
    }
    case TokenTypes.Paren: {
      ({ node, current } = readBarStartedExpressions(tokens, current))
      break
    }
    case TokenTypes.OperatorSup:
    case TokenTypes.OperatorA: {
      // high priority, do not look forward
      ({ node, current } = getParamOneNode(tokens, current, false))
      break
    }
    case TokenTypes.OperatorMinus: {
      // low priority, do look forward
      ({ node, current } = getParamOneNode(tokens, current, true))
      break
    }
    case TokenTypes.OperatorOAB: {
      node = createParamTwoNode()
      node.tex = token.tex
      current++
      const param0 = walk(tokens, current)
      current = param0.current
      if (param0.node.type === NodeTypes.Flat)
        param0.node = removeParenOfFlatExpr(param0.node)
      node.params[0] = param0.node
      const param1 = walk(tokens, current)
      current = param1.current
      if (param1.node.type === NodeTypes.Flat)
        param1.node = removeParenOfFlatExpr(param1.node)
      node.params[1] = param1.node
      break
    }
    case TokenTypes.OperatorO2: {
      current++
      if (current >= tokens.length) {
        node = createConstNode(`${token.tex.replace(/[\{\[] \$\d+ [\}\]]/g, '')}{}`)
        break
      }

      // detect next token
      ({ node, current } = lookForwardOperatorOptionalTwoParams(tokens, current, token))
      break
    }
    case TokenTypes.Split:
    case TokenTypes.Align: {
      current++
      node = createConstNode(token)
      break
    }
    case TokenTypes.RParen: {
      current++
      node = createConstNode(token)
      break
    }
    default: {
      throw new Error(`Unmatched token in walk ${token.value}`)
    }
  }
  // watch next token
  if (current < tokens.length && watchNext) {
    let matched = true
    while (matched && current < tokens.length) {
      const nextToken = tokens[current]
      switch (nextToken.type) {
        case TokenTypes.OperatorAOB: {
          ({ node, current } = preProcessOperatorAOB(node, nextToken, tokens, current + 1))
          break
        }
        case TokenTypes.OperatorAO: {
          node = preProcessOperatorAO(node, nextToken)
          current++
          break
        }
        case TokenTypes.OperatorSup: {
          ({ node, current } = preProcessOperatorSup(node, nextToken, tokens, current + 1))
          break
        }
        default: {
          matched = false
        }
      }
    }
  }
  return { node, current }
}

function parser(tokens: TokenizedValue[]) {
  const root = createRootNode()
  let current = 0

  while (current < tokens.length) {
    const walkRes = walk(tokens, current)
    current = walkRes.current
    root.body.push(walkRes.node)
  }

  return root
}

export {
  NodeTypes,
  Node,
  ChildNode,
  RootNode,
  AlignDirection,
  MatrixNode,
  parser,
}
