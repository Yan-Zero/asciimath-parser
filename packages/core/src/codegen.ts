import type { ChildNode, MatrixNode, RootNode } from './parser'
import { AlignDirection, NodeTypes } from './parser'

function getArrayBoundary(node: MatrixNode) {
  const div = node.dividerIndices
  let beginArray = '\\begin{array}'
  let ch = 'c'
  if (node.alignment === AlignDirection.Left)
    ch = 'l'

  if (div.length) {
    const divMax = div[div.length - 1]
    for (let i = div.length - 1; i >= 1; i--)
      div[i] -= div[i - 1]

    beginArray += '{'
    for (let i = 0; i < div.length; i++)
      beginArray += `${''.padEnd(div[i], ch)}|`

    // MathJax would complain if the array env arg
    // is not consistent with the elements of the matrix.
    const maxCol = Math.max(...node.params.map(i => i.length))
    beginArray += `${''.padEnd(maxCol - divMax, ch)}}`
  }
  else {
    const maxCol = Math.max(...node.params.map(c => c.length))
    beginArray += '{' + ''.padEnd(maxCol, ch) + '}'
  }
  return [beginArray, '\\end{array}']
}

function codegen(node: ChildNode | RootNode): string {
  switch (node.type) {
    case NodeTypes.Const: {
      return node.tex
    }
    case NodeTypes.Root: {
      let res = node.body.map(codegen).join(' ')
      if (node.body.find(n => n.type === NodeTypes.Const && n.value === '&'))
        res = `\\begin{aligned}${res}\\end{aligned}`
      return res
    }
    case NodeTypes.Flat: {
      return node.body.map(codegen).join(' ')
    }
    case NodeTypes.Matrix: {
      // const [beginMatrix, endMatrix] = getMatrixBoundary(node)
      const [arrayBegin, arrayEnd] = getArrayBoundary(node)
      return [
        node.lparen,
        // beginMatrix,
        arrayBegin,
        node.params.map(i => i.map(codegen).join('&')).join('\\\\'),
        '\\\\',
        arrayEnd,
        // endMatrix,
        node.rparen,
      ].join(' ')
    }
    case NodeTypes.ParamOne: {
      return node.tex.replace('$1', codegen(node.params))
    }
    case NodeTypes.ParamTwo: {
      return node.tex.replace('$1', codegen(node.params[0])).replace('$2', codegen(node.params[1]))
    }
  }
}

export {
  codegen,
}
