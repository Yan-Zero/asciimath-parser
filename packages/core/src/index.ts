import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
      ['#', '\\displaystyle'],
    ],
    replaceBeforeTokenizing: [
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (typeof config?.display !== 'undefined')
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      return `\\text{${String(e)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
