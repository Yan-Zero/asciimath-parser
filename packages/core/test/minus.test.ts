import { describe, expect, it } from 'vitest'
import { createTrie } from '../src/trie'
import { parser } from '../src/parser'
import { codegen } from '../src/codegen'
import { AsciiMath } from '../src'

describe('minus edge cases', () => {
//   const am = new AsciiMath({ display: false })
  it('should parse `e^-x`', () => {
    const code = 'e^-x'
    const trie = createTrie()
    const tokens = trie.tryParsingAll(code)
    expect(tokens).toMatchSnapshot()
    const ast = parser(tokens)
    expect(ast).toMatchSnapshot()
    const res = codegen(ast)
    expect(res).toMatchSnapshot()
  })

  it('should parse int_(0_-)', () => {
    const code = 'int_(0_-) (asd)'
    const trie = createTrie()
    const tokens = trie.tryParsingAll(code)
    expect(tokens).toMatchSnapshot()
    const ast = parser(tokens)
    expect(ast).toMatchSnapshot()
    const res = codegen(ast)
    expect(res).toMatchSnapshot()
  })

  it('should parse common minus expressions', () => {
    const am = new AsciiMath({ display: false })
    expect(am.toTex('a - b + c')).toMatchSnapshot()
    expect(am.toTex('a - b - c')).toMatchSnapshot()
  })
})

describe('minus with other operators', () => {
  const am = new AsciiMath({ display: false })
  it('should parse div', () => {
    expect(am.toTex('a/-b')).toMatchInlineSnapshot('"\\\\frac{ a }{ {-b } }"')
    expect(am.toTex('-a/b')).toMatchInlineSnapshot('"- \\\\frac{ a }{ b }"')
    expect(am.toTex('c/d-a/b')).toMatchInlineSnapshot('"\\\\frac{ c }{ d } - \\\\frac{ a }{ b }"')
  })

  it('should parse abs', () => {
    expect(am.toTex('abs -x')).toMatchInlineSnapshot('"\\\\left|{-x }\\\\right|"')
  })

  it('should parse integral', () => {
    expect(am.toTex('int_0^+oo')).toMatchInlineSnapshot('"\\\\int _{ 0 } ^{ {+\\\\infty } }"')
  })
})
