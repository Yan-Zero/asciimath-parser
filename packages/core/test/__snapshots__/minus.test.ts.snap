// Vitest Snapshot v1

exports[`minus edge cases > should parse \`e^-x\` 1`] = `
[
  {
    "current": 1,
    "isKeyWord": false,
    "tex": "e",
    "type": "StringLiteral",
    "value": "e",
  },
  {
    "current": 2,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 3,
    "isKeyWord": true,
    "tex": "{-$1 }",
    "type": "OperatorMinus",
    "value": "-",
  },
  {
    "current": 4,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
]
`;

exports[`minus edge cases > should parse \`e^-x\` 2`] = `
{
  "body": [
    {
      "body": [
        {
          "tex": "e",
          "type": "Const",
          "value": "e",
        },
        {
          "params": {
            "params": {
              "tex": "x",
              "type": "Const",
              "value": "x",
            },
            "tex": "{-$1 }",
            "type": "ParamOne",
          },
          "tex": "^{ $1 }",
          "type": "ParamOne",
        },
      ],
      "type": "Flat",
    },
  ],
  "type": "Root",
}
`;

exports[`minus edge cases > should parse \`e^-x\` 3`] = `"e ^{ {-x } }"`;

exports[`minus edge cases > should parse common minus expressions 1`] = `"a - b + c"`;

exports[`minus edge cases > should parse common minus expressions 2`] = `"a - b - c"`;

exports[`minus edge cases > should parse int_(0_-) 1`] = `
[
  {
    "current": 3,
    "isKeyWord": true,
    "tex": "\\\\int",
    "type": "Const",
    "value": "int",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 6,
    "isKeyWord": false,
    "tex": "0",
    "type": "NumberLiteral",
    "value": "0",
  },
  {
    "current": 7,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorSup",
    "value": "_",
  },
  {
    "current": 8,
    "isKeyWord": true,
    "tex": "{-$1 }",
    "type": "OperatorMinus",
    "value": "-",
  },
  {
    "current": 9,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 13,
    "isKeyWord": false,
    "tex": "as",
    "type": "StringLiteral",
    "value": "as",
  },
  {
    "current": 14,
    "isKeyWord": false,
    "tex": "d",
    "type": "StringLiteral",
    "value": "d",
  },
  {
    "current": 15,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;

exports[`minus edge cases > should parse int_(0_-) 2`] = `
{
  "body": [
    {
      "body": [
        {
          "tex": "\\\\int",
          "type": "Const",
          "value": "int",
        },
        {
          "params": {
            "body": [
              {
                "body": [
                  {
                    "tex": "0",
                    "type": "Const",
                    "value": "0",
                  },
                  {
                    "params": {
                      "tex": "-",
                      "type": "Const",
                      "value": "-",
                    },
                    "tex": "_{ $1 }",
                    "type": "ParamOne",
                  },
                ],
                "type": "Flat",
              },
            ],
            "type": "Flat",
          },
          "tex": "_{ $1 }",
          "type": "ParamOne",
        },
      ],
      "type": "Flat",
    },
    {
      "body": [
        {
          "tex": "\\\\left(",
          "type": "Const",
          "value": "(",
        },
        {
          "tex": "as",
          "type": "Const",
          "value": "as",
        },
        {
          "tex": "d",
          "type": "Const",
          "value": "d",
        },
        {
          "tex": "\\\\right)",
          "type": "Const",
          "value": ")",
        },
      ],
      "type": "Flat",
    },
  ],
  "type": "Root",
}
`;

exports[`minus edge cases > should parse int_(0_-) 3`] = `"\\\\int _{ 0 _{ - } } \\\\left( as d \\\\right)"`;

exports[`superscript with parens > should generate "e"^-(x+y) 1`] = `
[
  {
    "current": 3,
    "isKeyWord": false,
    "tex": "e",
    "type": "Text",
    "value": "e",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "{-$1 }",
    "type": "OperatorMinus",
    "value": "-",
  },
  {
    "current": 6,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 8,
    "isKeyWord": true,
    "tex": "{+$1 }",
    "type": "OperatorMinus",
    "value": "+",
  },
  {
    "current": 9,
    "isKeyWord": false,
    "tex": "y",
    "type": "StringLiteral",
    "value": "y",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;

exports[`superscript with parens > should generate "e"^-(x+y) 2`] = `
{
  "body": [
    {
      "body": [
        {
          "tex": "\\\\text{e}",
          "type": "Const",
          "value": "e",
        },
        {
          "params": {
            "params": {
              "body": [
                {
                  "tex": "\\\\left(",
                  "type": "Const",
                  "value": "(",
                },
                {
                  "tex": "x",
                  "type": "Const",
                  "value": "x",
                },
                {
                  "tex": "+",
                  "type": "Const",
                  "value": "+",
                },
                {
                  "tex": "y",
                  "type": "Const",
                  "value": "y",
                },
                {
                  "tex": "\\\\right)",
                  "type": "Const",
                  "value": ")",
                },
              ],
              "type": "Flat",
            },
            "tex": "{-$1 }",
            "type": "ParamOne",
          },
          "tex": "^{ $1 }",
          "type": "ParamOne",
        },
      ],
      "type": "Flat",
    },
  ],
  "type": "Root",
}
`;
